### PR TITLE
fix: Issue 166 fixed

### PR DESCRIPTION
**Pull request by yahu1031**
## Description
As per issue #166, the user wasn't able to continue without `netstat` installed on their machine. So I'm logging out a print statement as a warning to install netstat and try running the setup-table manually from the `project_server` dir.

### List of the issues solved

https://github.com/serverpod/serverpod/issues/166

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

## Breaking changes

_No breaking changes were made_

- Added new print statements like [`printWarning()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L33), [`printWarningln()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L36), [`printError()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L39), [`printErrorln()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L42) and changed `printww()` => [`_printww()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L7) and wrote another new [`printww()`](https://github.com/yahu1031/serverpod/blob/issue-166/tools/serverpod_cli/bin/util/print.dart#L24).

### CODE REVIEW

**Comment by yahu1031** [comment]
image
Sample output for the warning and error print statements

File: tools/serverpod_cli/pubspec.yaml [hunk 1]
```diff
@@ -19,7 +19,7 @@ dependencies:
   recase: ^4.0.0
   super_string: ^1.0.3
   watcher: ^1.0.0
-  analyzer: ^3.3.1
+  analyzer: ^4.0.0
```
Comment (vlidholt) [inline tools/serverpod_cli/pubspec.yaml:22]: This shouldn't be part of this PR

File: tools/serverpod_cli/bin/util/print.dart [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import 'dart:io';
 
+import 'package:colorize/colorize.dart';
 import 'package:super_string/super_string.dart';
 
-/// Print to terminal with word wrap.
-void printww(String string) {
+/// Private print function to terminal with word wrap.
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/util/print.dart:6]: Moving this to a private method serves no purpose?

File: tools/serverpod_cli/bin/util/print.dart [hunk 3]
```diff
@@ -19,7 +20,23 @@ void printww(String string) {
...
+/// Print to terminal with word wrap.
+void printww(String string) => _printww(string);
+
+/// Print to terminal with a new line.
 void printwwln(String string) {
   printww(string);
   printww('');
 }
+
+/// Print to terminal as warning in yellow color.
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/util/print.dart:32]: I will not accept this change unless it is used consistently throughout the CLI tooling. It should also be a separate PR.

File: tools/serverpod_cli/bin/util/command_line_tools.dart [hunk 4]
```diff
@@ -45,6 +45,12 @@ class CommandLineTools {
 
   static Future<void> createTables(Directory dir, String name) async {
     var serverPath = p.join(dir.path, '${name}_server');
+    var _isNetstatAvailable = await existsCommand('netstat');
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/util/command_line_tools.dart:48]: This part looks good.

File: tools/serverpod_cli/bin/util/command_line_tools.dart [hunk 5]
```diff
@@ -45,6 +45,12 @@ class CommandLineTools {
...
+    if (!_isNetstatAvailable) {
+      printWarning(
+          'netstat is not available on this platform. Please install it and tr running setup-tables manually from $serverPath.');
```
Comment (vlidholt) [inline tools/serverpod_cli/bin/util/command_line_tools.dart:51]: Please check spelling. Also, "not available on this platform" implies that it's not possible to install, so the error message needs better wording.

Also, is this tested with both netstat installed and not installed?

**Review by vlidholt — CHANGES_REQUESTED** [review]
Please see the change requests.

File: tools/serverpod_cli/bin/util/command_line_tools.dart (hunk 5 above)
Comment (yahu1031) [reply tools/serverpod_cli/bin/util/command_line_tools.dart:51]: I did it manually on the PC and will check that on non installed part. Anyhow cli tests are passing `existsCommand('xxxxx')` function. So considering that will work with non-installed platforms too. I will check with the script on installed part and update you in few minutes.

File: tools/serverpod_cli/bin/util/print.dart (hunk 3 above)
Comment (yahu1031) [reply tools/serverpod_cli/bin/util/print.dart:32]: Yes will make another PR for this part

File: tools/serverpod_cli/bin/util/command_line_tools.dart (hunk 5 above)
Comment (yahu1031) [reply tools/serverpod_cli/bin/util/command_line_tools.dart:51]: `netstat is not installed on this machine. Aborting tables creation. Please install it and run setup-tables script manually for tables creation.`

Does this sound good?

Yes will update the status on non-installed part soon in minutes

File: tools/serverpod_cli/pubspec.yaml (hunk 1 above)
Comment (yahu1031) [reply tools/serverpod_cli/pubspec.yaml:22]: Will fix it

File: tools/serverpod_cli/bin/util/print.dart (hunk 2 above)
Comment (yahu1031) [reply tools/serverpod_cli/bin/util/print.dart:6]: reverted.

File: tools/serverpod_cli/bin/util/command_line_tools.dart (hunk 5 above)
Comment (yahu1031) [reply tools/serverpod_cli/bin/util/command_line_tools.dart]: ### The machine which doesn't have `netstat` 

image

### The machine which has `netstat` 

image

**Comment by vlidholt** [comment]
I'm on the go now. I will need to test these changes out before I can merge it, so it may take a few days.

**Comment by vlidholt** [comment]
I had a closer look. I think the real fix to #166 isn't to prompt the user to install `netstat`, but instead find a way to make it work with some other tool. There must be something equivalent installed on Linux systems by default or possibly use ffi to perform the check.

**Comment by yahu1031** [comment]
Sounds like `netstat` is obsolete in Linux now and was replaced with `socket statistics(ss)`. Will need to change the code and will push it.

**Comment by yahu1031** [comment]
Here is some info about it [deprecated docs](https://dougvitale.wordpress.com/2011/12/21/deprecated-linux-networking-commands-and-their-replacements/)

**Comment by vlidholt** [comment]
Fixed by using ss when netstat is not available.
